triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala

```diff
@@ -0,0 +1,268 @@
+import com.auth0.jwt.algorithms.Algorithm
+import com.daml.auth.middleware.oauth2.Config.{
+  DefaultCookieSecure,
+  DefaultHttpPort,
+  DefaultLoginTimeout,
+  DefaultMaxLoginRequests,
+}
+import com.daml.cliopts
+import com.daml.jwt.{
+  ECDSAVerifier,
+  HMAC256Verifier,
+  JwksVerifier,
+  JwtVerifierBase,
+  JwtVerifierConfigurationCli,
+  RSA256Verifier,
+}
+import com.typesafe.scalalogging.StrictLogging
+import pureconfig.{ConfigReader, ConfigSource, ConvertHelpers}
+import pureconfig.error.ConfigReaderException
+import scopt.OptionParser
+
+import java.io.File
+import pureconfig.generic.semiauto._
+
+import java.nio.file.{Path, Paths}
+import scala.concurrent.duration
+import scala.concurrent.duration.FiniteDuration
+
+sealed trait ConfigError extends Product with Serializable {
+  def msg: String
+}
+case object MissingConfigError extends ConfigError {
+  val msg = "Missing auth middleware config file"
+}
+final case class ConfigParseError(msg: String) extends ConfigError
+
+case class Cli(
+    configFile: Option[File] = None,
+    // Host and port the middleware listens on
+    address: String = cliopts.Http.defaultAddress,
+    port: Int = DefaultHttpPort,
+    portFile: Option[Path] = None,
+    // The URI to which the OAuth2 server will redirect after a completed login flow.
+    // Must map to the `/cb` endpoint of the auth middleware.
+    callbackUri: Option[Uri] = None,
+    maxLoginRequests: Int = DefaultMaxLoginRequests,
+    loginTimeout: FiniteDuration = DefaultLoginTimeout,
+    cookieSecure: Boolean = DefaultCookieSecure,
+    // OAuth2 server endpoints
+    oauthAuth: Uri,
+    oauthToken: Uri,
+    // OAuth2 server request templates
+    oauthAuthTemplate: Option[Path],
+    oauthTokenTemplate: Option[Path],
+    oauthRefreshTemplate: Option[Path],
+    // OAuth2 client properties
+    clientId: String,
+    clientSecret: SecretString,
+    // Token verification
+    tokenVerifier: JwtVerifierBase,
+) {
+
+  import Cli._
+
+  def loadConfigFromFile: Either[ConfigError, Config] = {
+    require(configFile.nonEmpty, "Config file should be defined to load app config")
+    configFile
+      .map(f =>
+        try {
+          Right(ConfigSource.file(f).loadOrThrow[Config])
+        } catch {
+          case ex: ConfigReaderException[_] => Left(ConfigParseError(ex.failures.head.description))
+        }
+      )
+      .get
+  }
+
+  def loadConfigFromCliArgs: Config = {
+    val cfg = Config(
+      address,
+      port,
+      portFile,
+      callbackUri,
+      maxLoginRequests,
+      loginTimeout,
+      cookieSecure,
+      oauthAuth,
+      oauthToken,
+      oauthAuthTemplate,
+      oauthTokenTemplate,
+      oauthRefreshTemplate,
+      clientId,
+      clientSecret,
+      tokenVerifier,
+    )
+    cfg.validate
+    cfg
+  }
+}
+
+object Cli extends StrictLogging {
+  implicit val tokenVerifierReader =
+    ConfigReader.forProduct2[JwtVerifierBase, String, String]("type", "uri") { case (t, p) =>
```

Comment:
```suggestion
    ConfigReader.forProduct2("type", "uri") { (t: String, p: String) =>
```

Reply:
fails , so leaving it as is
```triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala:109: error: missing parameter type for expanded function
The argument types of an anonymous function must be fully known. (SLS 8.5)
Expected type was: (?, ?) => com.daml.jwt.JwtVerifierBase
    ConfigReader.forProduct2("type", "uri") { case (t: String, p: String) =>
 ```

Reply:
That's because you left the `case` keyword there, which is removed in my suggestion. This isn't a function on a 2-tuple argument, and `case (t: String, p: String)` _does not contain any type annotations_.

Reply:
Smaller example

```scala
scala> def f2[A, B, R](f: (A, B) => R) = f
def f2[A, B, R](f: (A, B) => R): (A, B) => R

scala> f2((x: Int, y: Int) => (x, y))
val res0: (Int, Int) => (Int, Int) = $Lambda$1028/2036295297@66d3b881

scala> f2{case (x: Int, y: Int) => (x, y)}
         ^
       error: missing parameter type for expanded function
       The argument types of an anonymous function must be fully known. (SLS 8.5)
       Expected type was: (?, ?) => ?
```